include/dem/pw_contact_force.h

```diff
@@ -47,13 +48,59 @@ class PWContactForce
    *
    * @param pw_pairs_in_contact Required information for calculation of the
    * particle-wall contact force
-   * @param dem_parameters DEM parameters declared in the .prm file
+   * @param physical_properties DEM physical properties declared in the .prm
+   * file
+   * @param dt DEM time step
    */
   virtual void
   calculate_pw_contact_force(
-    const std::map<int, std::map<int, pw_contact_info_struct<dim>>>
-      *                             pw_pairs_in_contact,
-    const DEMSolverParameters<dim> &dem_parameters) = 0;
+    std::unordered_map<int, std::map<int, pw_contact_info_struct<dim>>>
+      *                                               pw_pairs_in_contact,
+    const Parameters::Lagrangian::PhysicalProperties &physical_properties,
+    const double &                                    dt) = 0;
+
+protected:
+  /**
+   * Carries out updating the contact pair information for both non-linear and
+   * linear contact force calculations
+   *
+   * @param contact_pair_information Contact information of a particle-wall pair
+   * in neighborhood
+   * @param particle_properties Properties of particle in contact
+   * @param dt DEM time step
+   */
+  void
+  update_contact_information(
+    pw_contact_info_struct<dim> &  contact_pair_information,
+    const ArrayView<const double> &particle_properties,
+    const double &                 dt);
+
+  /**
+   * Carries out applying the calculated force and torque on the local-local
+   * particle pair in contact, for both non-linear and linear contact force
+   * calculations
+   *
+   * @param particle_one_properties Properties of particle one in contact
+   * @param particle_two_properties Properties of particle two in contact
+   * @param forces_and_torques A tuple which contains: 1, normal force, 2,
+   * tangential force, 3, tangential torque and 4, rolling resistance torque of
+   * a contact pair
+   */
+  void
+  apply_force_and_torque(ArrayView<double> &               particle_properties,
+                         const std::tuple<Tensor<1, dim>,
+                                          Tensor<1, dim>,
+                                          Tensor<1, dim>,
+                                          Tensor<1, dim>> &forces_and_torques);
+
+  /** This function is used to find the projection of vector_a on
+   * vector_b
+   * @param vector_a A vector which is going to be projected on vector_b
+   * @param vector_b The projection vector of vector_a
+   * @return The projection of vector_a on vector_b
+   */
+  Tensor<1, dim> find_projection(Tensor<1, dim> vector_a,
+                                 Tensor<1, dim> vector_b);
```

Comment:
This function does not modify vector_a and vector_b, so instead you could pass it by const reference instead of copy. This would prevent copy-creation of the tensors.